distributed/scheduler.py

```diff
@@ -5673,143 +5674,172 @@ async def gather(self, keys, serializers=None):
-            bad_nannies = [
-                addr for addr, resp in zip(nanny_workers, resps) if resp is not None
-            ]
-            if bad_nannies:
-                await asyncio.gather(
-                    *(
-                        self.remove_worker(addr, stimulus_id=stimulus_id)
-                        for addr in bad_nannies
+            bad_nannies = []
+            for nanny, response in zip(nanny_addresses, responses):
+                if isinstance(response, TimeoutError):
+                    logger.error(
+                        "Scheduler timed out trying to restart worker on %s.", nanny
                     )
-                )
-
-                raise TimeoutError(
-                    f"{len(bad_nannies)}/{len(nannies)} nanny worker(s) did not shut down within {timeout}s"
-                )
-
-        self.log_event([client, "all"], {"action": "restart", "client": client})
-
-        if wait_for_workers:
-            while len(self.workers) < n_workers:
-                # NOTE: if new (unrelated) workers join while we're waiting, we may return before
-                # our shut-down workers have come back up. That's fine; workers are interchangeable.
-                if monotonic() < start + timeout:
-                    await asyncio.sleep(0.2)
-                else:
-                    msg = (
-                        f"Waited for {n_workers} worker(s) to reconnect after restarting, "
-                        f"but after {timeout}s, only {len(self.workers)} have returned. "
-                        "Consider a longer timeout, or `wait_for_workers=False`."
+                    bad_nannies.append(nanny)
+                elif isinstance(response, Exception):
+                    logger.exception(
+                        "Encountered exception trying to to restart worker on %s.",
+                        nanny,
+                        exc_info=response,
                     )
-
-                    if (n_nanny := len(nanny_workers)) < n_workers:
-                        msg += (
-                            f" The {n_workers - n_nanny} worker(s) not using Nannies were just shut "
-                            "down instead of restarted (restart is only possible with Nannies). If "
-                            "your deployment system does not automatically re-launch terminated "
-                            "processes, then those workers will never come back, and `Client.restart` "
-                            "will always time out. Do not use `Client.restart` in that case."
-                        )
-                    raise TimeoutError(msg) from None
-        logger.info("Restarting finished.")
+                    bad_nannies.append(nanny)
+                elif response == "timed out":
+                    logger.error("%s timed out trying to restart its worker.", nanny)
+                    bad_nannies.append(nanny)
+            if bad_nannies:
+                raise RuntimeError(
```

Comment:
We may want to differentiate between "pure" timeouts, i.e. all the errors we see are a `TimeoutError` and raise a `TimeoutError` if that's the case and raise a generic `RuntimeError` as an aggregation of all occurring errors otherwise to signify that we observed an actual problem instead of just timing out. Then again, I'm not sure if that's really a problem and would warrant the additional logic.

Reply:
Another thought: The benefit of collecting errors and handling them in bulk is that we can pass more information to the user _if_ we return before the timeout. For example, if 5 nannies fail to restart and the other 95 succeed, we will be able to tell that to the user. The downside is that we do not pass any information back to the user if a single nanny times out.